chore: adding implementation for SendPacket message server

chatton opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

This PR adds the bulk of the implementation for the SendPacket message server V2 implementation. There are a few follow up items that need to be completed before this handler is complete. I will link these in follow up issues.


Part of: #7359

closes: #7353

<!-- Please refer to the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages in ibc-go.

This repository uses [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).

Example commit messages:

fix: skip emission of unpopulated memo field in ics20
deps: updating sdk to v0.46.4
chore: removed unused variables
e2e: adding e2e upgrade test for ibc-go/v6
docs: ics27 v6 documentation updates
feat: add semantic version utilities for e2e tests
feat(api)!: this is an api breaking feature
fix(statemachine)!: this is a statemachine breaking fix
-->

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

chatton commented on modules/core/04-channel/types/timeout.go:
```diff
@@ -14,6 +14,14 @@ func NewTimeout(height clienttypes.Height, timestamp uint64) Timeout {
 }
 
+// NewTimeoutWithTimestamp creates a new Timeout with only the timestamp set.
+func NewTimeoutWithTimestamp(timestamp uint64) Timeout {
```
we only care about timestamp in the eureka spec

chatton commented on modules/core/04-channel/v2/keeper/packet.go:
```diff
@@ -0,0 +1,78 @@
+	timeoutTimestamp uint64,
+	data []channeltypesv2.PacketData,
+) (uint64, error) {
+	// TODO: add aliasing logic
```
@chatton  create an issue for this!

chatton commented on modules/core/04-channel/v2/types/expected_keepers.go:
```diff
@@ -0,0 +1,25 @@
+package types
```
duplicate from packet-server keeper

chatton commented on modules/core/05-port/types/module.go:
```diff
@@ -108,6 +108,28 @@ type IBCModule interface {
+	// OnSendPacket is executed when a packet is being sent from sending chain.
+	// this callback is provided with the source and destination IDs, the signer, the packet sequence and the packet data
+	// for this specific application.
+	OnSendPacket(
```
I figure we can add them one method at a time

chatton commented on modules/core/05-port/types/module.go:
```diff
@@ -108,6 +108,28 @@ type IBCModule interface {
 
+// IBCModuleV2 defines an interface that implements all the callbacks
+// that modules must define as specified in IBC Protocol V2
+type IBCModuleV2 interface {
```
not convinced this is the best place for this interface, open to suggestions! (Also do we want V2 in the name? )

colin-axner replied:
It really should be referenced as `api.IBCModule`, I wonder if it makes sense to create the api directory now, would be a weird partial transition if you didn't move over the light client module

Main dependency concern is that if you move this api elsewhere, you may need to define interfaces or move the concrete implementations

chatton commented on modules/core/04-channel/v2/types/errors.go:
```diff
@@ -0,0 +1,10 @@
+package types
```
duplicate from packet-server

chatton commented on modules/core/04-channel/v2/keeper/packet.go:
```diff
@@ -0,0 +1,78 @@
+	errorsmod "cosmossdk.io/errors"
+
+	clienttypes "github.com/cosmos/ibc-go/v9/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/v9/modules/core/04-channel/types"
```
Maybe we want to a new timeout type that lives just in channeltypesv2? To not need this dependency

colin-axner replied:
the timeout being referenced by the light client interface seems unavoidable to me, so I think that's the main consideration. Checkout https://github.com/cosmos/ibc-go/issues/7055 for context on how this might look going forward as well. In #7055, you could remove the timeout type and separate legacy behaviour by providing the timestamp/height type separately

colin-axner commented on modules/core/04-channel/v2/keeper/msg_server.go:
```diff
@@ -0,0 +1,53 @@
+
+	signer, err := sdk.AccAddressFromBech32(msg.Signer)
+	if err != nil {
+		sdkCtx.Logger().Error("send packet failed", "error", errorsmod.Wrap(err, "Invalid address for msg Signer"))
+		return nil, errorsmod.Wrap(err, "Invalid address for msg Signer")
```
```suggestion
		sdkCtx.Logger().Error("send packet failed", "error", errorsmod.Wrap(err, "invalid address for msg Signer"))
		return nil, errorsmod.Wrap(err, "invalid address for msg Signer")
```

chatton commented on modules/core/04-channel/v2/keeper/packet.go:
yeah I think we should call it relay.go

chatton commented:
Sorry guys! I ended up making a few more changes,

- I added the aliasing logic as suggested
- included some tests (wiring needed to happen to ibc core keeper to allow this)
- moved module into api dir (great suggestion @colin-axner )
